### PR TITLE
[GEOS-9834] GetLegendGraphic results in java.lang.NegativeArraySizeException or OutOfMemory

### PR DESCRIPTION
**Pull request by spike83**
According to my bug report [GEOS-9834](https://osgeo-org.atlassian.net/browse/GEOS-9834) I opened this pr to fix the scaling of the image.

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

**Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.**

For all pull requests:

- [X] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [X] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [X] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [X] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates (screenshots, text)
- [ ] Commits changing the REST API, or any configuration object, should check if the REST API docs (Swagger YAML files and classic documentation) need to be updated.

### CODE REVIEW

File: src/wms/src/main/java/org/geoserver/wms/legendgraphic/LegendGraphicBuilder.java [hunk 1]
```diff
@@ -153,8 +153,8 @@ protected Style resizeForDPI(GetLegendGraphicRequest request, Style gt2Style) {
         double standardDpi = RendererUtilities.getDpi(Collections.emptyMap());
         if (dpi != standardDpi) {
             double scaleFactor = dpi / standardDpi;
-            w = ((int) Math.round(w * scaleFactor));
-            h = ((int) Math.round(h * scaleFactor));
+            w = ((int) Math.round(request.getWidth() * scaleFactor));
+            h = ((int) Math.round(request.getHeight() * scaleFactor));
```
Comment (aaime) [inline src/wms/src/main/java/org/geoserver/wms/legendgraphic/LegendGraphicBuilder.java:157]: I am not familiar with this bit of code, but after review, the fix seems legit.
However, I'm not convinced it's the best way to solve this. The method is called over and over, once for each layer, the reason why we're having this problem it's that the code does two separate things:
* Resize the style based on the DPI, its intended purpose from the method name.
* As a side effect, also change the w and h fields... and if done over and over, well, that can cause very bad results indeed.

I would propose to move the side effect to another method that is called outside of the loop instead, e..g, ``resizeBoundsForDPI``. The call would be before the "for" loops in ``BufferedImageLegendGraphicBuilder`` and ``JSONLegendGraphicBuilder``. @ianturton you seem to be the original author of this method (as far as git goes at least), what do you think?

Comment (ianturton) [reply src/wms/src/main/java/org/geoserver/wms/legendgraphic/LegendGraphicBuilder.java:157]: I was just looking at that and thinking it looked familiar! 

Looks as if whatever I was working on only ever called that section of code once. So I would pull out the resize as @aaime suggests.

**Comment by mprins** [comment]
I restarted the build as it locked up on downloading something from the central maven repo

**Comment by aaime** [comment]
Looks good to me, merging! 
@spike83 if you want the fix also on the stable and maintenance branch, you'll have to make backports PRs.

**Pull request merged**